charms/worker/k8s/src/charm.py

```diff
@@ -234,12 +237,49 @@ def _create_cos_tokens(self):
     @on_error(
         WaitingStatus("Waiting for enable components"), InvalidResponseError, K8sdConnectionError
     )
-    def _enable_components(self):
+    def _configure_components(self):
         """Enable necessary components for the Kubernetes cluster."""
-        status.add(ops.MaintenanceStatus("Enabling DNS"))
-        self.api_manager.enable_component("dns", True)
+        if self._dns_charm_integrated():
+            status.add(ops.MaintenanceStatus("Disabling DNS"))
+            self.api_manager.configure_component("dns", False)
+            self._configure_dns()
+        else:
+            status.add(ops.MaintenanceStatus("Enabling DNS"))
+            self.api_manager.configure_component("dns", True)
+
         status.add(ops.MaintenanceStatus("Enabling Network"))
-        self.api_manager.enable_component("network", True)
+        self.api_manager.configure_component("network", True)
+
+    def _dns_charm_integrated(self) -> bool:
+        """Check if the DNS charm is integrated.
+
+        Returns:
+            bool: True if the DNS charm is integrated, False otherwise.
+        """
+        dns_relation = self.model.get_relation("dns-provider")
+        if not dns_relation:
+            return False
+        return True
+
+    def _configure_dns(self):
```

Comment:
pass that object in here -- it could be None
```suggestion
    def _configure_dns(self, settings: Optional[DNSSettings]):
```